compiler/frontend/src/mir_optimize/common_subtree_elimination.rs

```diff
@@ -56,24 +56,32 @@ pub fn eliminate_common_subtrees(body: &mut Body, pureness: &PurenessInsights) {
     // The matches are stored as an index into [body] so that we can read the
     // potentially matching normalized expression and also have mutable access
     // to the current expression within the main loop.
-    let mut pure_expressions: FxHashMap<u64, Vec<usize>> = FxHashMap::default();
+    let mut deterministic_expressions: FxHashMap<u64, Vec<usize>> = FxHashMap::default();
 
     let mut inner_function_ids: FxHashMap<Id, Vec<Id>> = FxHashMap::default();
     let mut additional_function_hirs: FxHashMap<Id, FxHashSet<hir::Id>> = FxHashMap::default();
-    let mut updated_references: FxHashMap<Id, Id> = FxHashMap::default();
+
+    // When two expressions are the same, the second one gets replaced by a
+    // reference to the first one. In order for common subtree elimination to
+    // work on expressions after that, which reference the second expression,
+    // we basically need to do reference following as well.
+    let mut replaced: FxHashMap<Id, Id> = FxHashMap::default();
 
     for index in 0..body.expressions.len() {
         let id = body.expressions[index].0;
 
-        let normalized_hash = {
-            let expression = &mut body.expressions[index].1;
-            expression.replace_id_references(&mut |id| {
-                if let Some(update) = updated_references.get(id) {
-                    *id = *update;
+        {
+            body.expressions[index].1.replace_ids(&mut |id| {
+                if let Some(other) = replaced.get(id) {
+                    *id = *other;
                 }
             });
+        }
```

Comment:
Now that this is only a single statement, we don't need the containing body anymore

```suggestion
        body.expressions[index].1.replace_ids(&mut |id| {
            if let Some(other) = replaced.get(id) {
                *id = *other;
            }
        });
```